garage/tf/policies/categorical_conv_policy_with_model.py

```diff
@@ -0,0 +1,131 @@
+        conv_strides: The stride of the sliding window.
+        conv_pad: The type of padding algorithm to use,
+            either 'SAME' or 'VALID'.
+        name: Variable scope of the policy.
+        hidden_sizes: Output dimension of dense layer(s).
+        hidden_nonlinearity: Activation function for
+                    intermediate dense layer(s).
+        output_nonlinearity: Activation function for
+                    output dense layer.
+        layer_normalization: Bool for using layer normalization or not.
+
+    """
+
+    def __init__(self,
+                 env_spec,
+                 conv_filters,
+                 conv_filter_sizes,
+                 conv_strides,
+                 conv_pad,
+                 name='CategoricalConvPolicy',
+                 hidden_sizes=[],
+                 hidden_nonlinearity=tf.nn.relu,
+                 output_nonlinearity=tf.nn.softmax,
+                 layer_normalization=False):
+        assert isinstance(env_spec.action_space, Discrete), (
+            'CategoricalConvPolicy only works with akro.tf.Discrete'
+            'action space.')
+        super().__init__(name, env_spec)
+        self.obs_dim = env_spec.observation_space.flat_dim
+        self.action_dim = env_spec.action_space.n
+
+        self.add_model(
+            CNNModel(
+                filter_dims=conv_filter_sizes,
+                num_filters=conv_filters,
+                strides=conv_strides,
+                padding=conv_pad,
+                hidden_nonlinearity=hidden_nonlinearity,
+                name='CNNModel'))
+        self.add_model(
+            MLPModel(
+                output_dim=self.action_dim,
+                hidden_sizes=hidden_sizes,
+                hidden_nonlinearity=hidden_nonlinearity,
+                output_nonlinearity=output_nonlinearity,
+                layer_normalization=layer_normalization,
+                name='MLPModel'))
+
+        self._initialize()
+
+    def _initialize(self):
+        state_input = tf.placeholder(tf.float32, shape=(None, self.obs_dim))
+
+        with tf.variable_scope(self._variable_scope):
+            self.build_models(state_input)
+
+        self._f_prob = tf.get_default_session().make_callable(
+            self.outputs, feed_list=[self.input])
+
+    @property
+    def vectorized(self):
+        """Vectorized or not."""
+        return True
+
+    @overrides
+    def dist_info_sym(self, obs_var, state_info_vars=None, name=None):
+        """Symbolic graph of the distribution."""
+        with tf.variable_scope(self._variable_scope):
+            prob = self.build_models(obs_var, name=name)
+        return dict(prob=prob)
+
+    @overrides
+    def dist_info(self, obs, state_infos=None):
+        """Distribution info."""
+        prob = self._f_prob(obs)
+        return dict(prob=prob)
+
+    @overrides
+    def get_action(self, observation):
+        """Return a single action."""
+        flat_obs = self.observation_space.flatten(observation)
```

Comment:
currently it flatten the 2D image, since we are doing `self.obs_dim = env_spec.observation_space.flat_dim` in the policy.
I actually think this is a mistake, it should not flatten the observation here. For example, in pixel environment we want to pass the original image input with shape (w, h, c) to the policy.
This is missed because the CNNModel was mocked out.

Reply:
cam you fix it?
i don't think the image should be flattened. that seems wrong.